doom.d/config.org
#+TITLE: Emacs Config
#+STARTUP: fold

* Startup
#+BEGIN_SRC emacs-lisp :comments no
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-
#+END_SRC
* Personal Info
Some functionality uses this to identify you, e.g. GPG configuration, email
clients, file templates and snippets.

#+BEGIN_SRC emacs-lisp
(setq user-full-name "Abhishek Bansal"
      user-mail-address "abhibansal530@gmail.com")
#+END_SRC
* Theme and look
** Font
#+BEGIN_SRC emacs-lisp
(setq doom-font (font-spec :family "Meslo LG S for Powerline" :size 18))
#+END_SRC
** Theme
#+BEGIN_SRC emacs-lisp
(setq doom-theme 'gruvbox-dark-soft)
#+END_SRC
** Misc
This determines the style of line numbers in effect. If set to `nil', line
numbers are disabled. For relative line numbers, set this to `relative'.
#+BEGIN_SRC emacs-lisp
(setq display-line-numbers-type t)
#+END_SRC
* Org mode
** Defaults
If you use `org' and don't want your org files in the default location below,
change `org-directory'. It must be set before org loads!
#+BEGIN_SRC emacs-lisp
(setq org-directory "~/org/")
#+END_SRC

** Main
#+BEGIN_SRC emacs-lisp
(after! org
  (require 'find-lisp)
  (setq org-ditaa-jar-path "/usr/local/Cellar/ditaa/0.11.0_1/libexec/ditaa-0.11.0-standalone.jar")
  (setq org-startup-folded t)
  (setq org-todo-keywords
        '((sequence "TODO(t)" "NEXT(n)" "SOMEDAY(s)" "PROJECT(p)" "|" "DONE(d)")
          (sequence "WAITING(w@/!)" "HOLD(h@/!)" "|" "CANCELLED(c@/!)")))

  ;; Alghout all of these are not required for agenda, but adding these allow for searching tags across all these files.
  (setq org-agenda-files (append '("~/org/gtd/inbox.org" "~/org/gtd/life.org" "~/org/gtd/projects.org")
                                 (directory-files-recursively "~/org/resources" "\\.org$")))

  (setq org-complete-tags-always-offer-all-agenda-tags t)

  (setq org-refile-allow-creating-parent-nodes 'confirm)

  (setq org-refile-targets '(("~/org/gtd/projects.org" :maxlevel . 1)
                             ("~/org/gtd/life.org" :maxlevel . 3))))
#+END_SRC

** Capture
#+BEGIN_SRC emacs-lisp
(after! org-capture
  (setq org-capture-templates
        `(("i" "Inbox")

          ("iP" "Personal inbox" entry
           (file "~/org/gtd/inbox.org")
           (file "~/org/capture-templates/inbox.txt")
           :empty-lines-after 1)

          ("iW" "Work inbox" entry
           (file "~/org/gtd/work/inbox.org")
           (file "~/org/capture-templates/inbox.txt")
           :empty-lines-after 1)

          ("c" "org-protocol-capture" entry
           (file "~/org/gtd/inbox.org")
           (file "~/org/capture-templates/protocol.txt")
           :immediate-finish t)

          ("g" "GTD templates")

          ("gp" "Create a daily plan")

          ("gpP" "Daily plan private" plain
           (file+olp+datetree "~/org/gtd/planner.org")
           (file "~/org/capture-templates/daily-plan.txt")
           :immediate-finish t)

          ("gpW" "Daily plan work" plain
           (file+olp+datetree "~/org/gtd/work/planner.org")
           (file "~/org/capture-templates/daily-plan-work.txt")
           :immediate-finish t)

          ("gr" "Review templates")

          ("grd" "Daily review" plain
           (file+olp+datetree "~/org/gtd/life.org" "GTD" "Daily Review")
           (file "~/org/capture-templates/daily-review.txt")
           :immediate-finish t
           :jump-to-captured t)

          ("grw" "Weekly review" plain
           (file+olp+datetree "~/org/gtd/life.org" "GTD" "Weekly Review")
           (file "~/org/capture-templates/weekly-review.txt")
           :immediate-finish t
           :jump-to-captured t)

          ("r" "Capture a resource")

          ("rB" "Book" entry
           (file "~/org/resources/books.org")
           (file "~/org/capture-templates/book.txt")
           :empty-lines-after 2))))
#+END_SRC
** Protocol
Setup org-protocol to enable capture using firefox
#+BEGIN_SRC emacs-lisp
(add-to-list 'load-path "~/.emacs.d/.local/straight/repos/org-mode/lisp/org-protocol.el")
(require 'org-protocol)
#+END_SRC
** Agenda
Define custom agenda views.
#+BEGIN_SRC emacs-lisp
(use-package! org-agenda
  :init
  :config
  (setq abhi/org-agenda-directory "~/org/gtd/")
  (setq org-columns-default-format "%40ITEM(Task) %Effort(EE){:} %CLOCKSUM(Time Spent) %SCHEDULED(Scheduled) %DEADLINE(Deadline)")
  (setq org-agenda-custom-commands `(("c" "Agenda"
                                      ((agenda ""
                                               ((org-agenda-span 'week)
                                                (org-deadline-warning-days 365)))
                                       (todo "TODO"
                                             ((org-agenda-overriding-header "To Refile")
                                              (org-agenda-files '(,(concat abhi/org-agenda-directory "inbox.org")))))
                                       (todo "NEXT"
                                             ((org-agenda-overriding-header "In Progress")
                                              (org-agenda-files '(,(concat abhi/org-agenda-directory "projects.org")
                                                                  ,(concat abhi/org-agenda-directory "next.org")))
                                              ))
                                       (todo "TODO"
                                             ((org-agenda-overriding-header "Projects")
                                              (org-agenda-files '(,(concat abhi/org-agenda-directory "projects.org")))
                                              ))
                                       (todo "TODO"
                                             ((org-agenda-overriding-header "One-off Tasks")
                                              (org-agenda-files '(,(concat abhi/org-agenda-directory "next.org")))
                                              (org-agenda-skip-function '(org-agenda-skip-entry-if 'deadline 'scheduled))))))

                                     ("i" "Inbox"
                                      ((todo ""
                                            ((org-agenda-overriding-header "To Refile")
                                             (org-agenda-files '(,(concat abhi/org-agenda-directory "inbox.org"))))))))))
#+END_SRC
** Roam
#+BEGIN_SRC emacs-lisp
(use-package! org-roam
  :hook
  (after-init . org-roam-mode)
  :config
  (setq org-roam-ref-capture-templates
        '(("r" "ref" plain (function org-roam-capture--get-point)
           "%?"
           :file-name "${slug}"
           :head "#+TITLE: ${title}
    #+ROAM_KEY: ${ref}
    #+ROAM_TAGS: Website
    - source :: ${ref}"
           :unnarrowed t)))
  (setq +org-roam-open-buffer-on-find-file nil))
#+END_SRC
** Journal
Store journal entries inside roam directory.
#+BEGIN_SRC emacs-lisp
(use-package org-journal
  :custom
  (org-journal-dir (concat org-roam-directory "/journal"))
  (org-journal-date-prefix "#+TITLE: ")
  (org-journal-file-format "%Y-%m-%d.org")
  (org-journal-date-format "%A, %d %B %Y"))
#+END_SRC
** Refile
My custom hydra for refiling. Ref :
- http://www.howardism.org/Technical/Emacs/getting-more-boxes-done.html
- https://mollermara.com/blog/Fast-refiling-in-org-mode-with-hydras/

#+BEGIN_SRC emacs-lisp
(defmacro my-org-make-refile-command (fn-suffix refile-targets)
  "Generate a command to call `org-refile' with modified targets."
  `(defun ,(intern (concat "my-org-refile-" (symbol-name fn-suffix))) ()
     ,(format "`org-refile' to %S" refile-targets)
     (interactive)
     (org-refile-cache-clear)
     (let ((org-refile-target-verify-function nil)
           (org-refile-history nil)
           (org-refile-targets ,refile-targets))
       (if (eq major-mode 'org-agenda-mode)
           (call-interactively 'org-agenda-refile)
         (call-interactively 'org-refile)))))

(my-org-make-refile-command book '(("~/org/resources/books.org" :maxlevel . 2)))
(my-org-make-refile-command course '(("~/org/resources/courses.org" :maxlevel . 1)))
(my-org-make-refile-command link '(("~/org/resources/links.org" :maxlevel . 1)))
(my-org-make-refile-command paper '(("~/org/resources/papers.org" :maxlevel . 1)))
(my-org-make-refile-command read '(("~/org/resources/readings.org" :maxlevel . 1)))
(my-org-make-refile-command watch '(("~/org/resources/watch.org" :maxlevel . 1)))
(my-org-make-refile-command project '(("~/org/gtd/projects.org" :maxlevel . 1)))
(my-org-make-refile-command people '(("~/org/resources/people.org" :maxlevel . 1)))

(defhydra my-org-refile-hydra (:hint nil :foreign-keys run)
  "
^Refile^            ^Goto^                     ^Dired^
------------------------------------------------------
_p_: Projects       _g j_: Last refile         _d r_: Resources
_P_: Papers         _g r_: To Read
_b_: Books          _g p_: Projects
_c_: Courses
_l_: Links
_r_: To Read
_w_: To Watch
_o_: People
_R_: Refile
"
  ("b" my-org-refile-book)
  ("c" my-org-refile-course)
  ("l" my-org-refile-link)
  ("p" my-org-refile-project)
  ("P" my-org-refile-paper)
  ("r" my-org-refile-read)
  ("w" my-org-refile-watch)
  ("o" my-org-refile-people)
  ("R" org-refile)
  ("g j" org-refile-goto-last-stored :exit t)
  ("g r" (find-file-other-window "~/org/resources/readings.org") :exit t)
  ("g p" (find-file-other-window "~/org/gtd/projects.org") :exit t)
  ("d r" (dired "~/org/resources") :exit t)
  ("q" nil "cancel"))

(global-set-key (kbd "<f9> r") 'my-org-refile-hydra/body)
#+END_SRC
** Misc
Use deft for searching org files :
#+BEGIN_SRC emacs-lisp
(use-package deft
      :after org
      :custom
      (deft-recursive t)
      (deft-use-filter-string-for-filename t)
      (deft-default-extension "org")
      (deft-directory org-roam-directory))
#+END_SRC

Don't rememer this one :

#+BEGIN_SRC emacs-lisp
(eval-after-load "artist"
  '(define-key artist-mode-map [(down-mouse-3)] 'artist-mouse-choose-operation)
  )
#+END_SRC
* Applications
** Ebook reading
For reading epubs :
#+BEGIN_SRC emacs-lisp
(use-package! nov
  :mode ("\\.epub\\'" . nov-mode)
  :config
  (setq nov-save-place-file (concat doom-cache-dir "nov-places")))
#+END_SRC
** RSS
*** Config
Path to feed file.
#+BEGIN_SRC emacs-lisp
(setq rmh-elfeed-org-files '("~/org/elfeed.org"))
#+END_SRC
*** Keybindings
#+BEGIN_SRC emacs-lisp
(map! :map elfeed-search-mode-map
      :after elfeed-search
      :n "e" #'elfeed-update
      :n "r" #'elfeed-search-untag-all-unread
      :n "u" #'elfeed-search-tag-all-unread)
#+END_SRC
*** Visual enhancements
These are taken from https://github.com/tecosaur/emacs-config/blob/master/config.org#visual-enhancements
#+BEGIN_SRC emacs-lisp
(after! elfeed

  (elfeed-org)

  (setq elfeed-search-filter "@1-week-ago +unread"
        elfeed-search-print-entry-function '+myrss/elfeed-search-print-entry
        elfeed-show-refresh-function #'+myrss/elfeed-show-refresh--better-style
        elfeed-search-title-min-width 80
        elfeed-search-title-max-width 100
        shr-max-image-proportion 0.6)

  (add-hook! 'elfeed-show-mode-hook (hide-mode-line-mode 1))
  (add-hook! 'elfeed-search-update-hook #'hide-mode-line-mode)

  (defface elfeed-show-title-face '((t (:weight ultrabold :slant italic :height 1.5)))
    "title face in elfeed show buffer"
    :group 'elfeed)
  (defface elfeed-show-author-face `((t (:weight light)))
    "title face in elfeed show buffer"
    :group 'elfeed)
  (set-face-attribute 'elfeed-search-title-face nil
                      :foreground 'nil
                      :weight 'light)

  (defadvice! +rss-elfeed-wrap-h-nicer ()
    "Enhances an elfeed entry's readability by wrapping it to a width of
`fill-column' and centering it with `visual-fill-column-mode'."
    :override #'+rss-elfeed-wrap-h
    (setq-local truncate-lines nil
                shr-width 120
                visual-fill-column-center-text t
                default-text-properties '(line-height 1.1))
    (let ((inhibit-read-only t)
          (inhibit-modification-hooks t))
      (visual-fill-column-mode)
      ;; (setq-local shr-current-font '(:family "Merriweather" :height 1.2))
      (set-buffer-modified-p nil)))

  ;; This defines how feed entries will be displayed.
  ;; <Feed Name>   <Entry Title>  <Tags>
  ;; TODO : Fix tags column
  (defun +myrss/elfeed-search-print-entry (entry)
    "Print ENTRY to the buffer."
    (let* ((elfeed-goodies/tag-column-width 40)
           (elfeed-goodies/feed-source-column-width 30)
           (title (or (elfeed-meta entry :title) (elfeed-entry-title entry) ""))
           (title-faces (elfeed-search--faces (elfeed-entry-tags entry)))
           (feed (elfeed-entry-feed entry))
           (feed-title
            (when feed
              (or (elfeed-meta feed :title) (elfeed-feed-title feed))))
           (tags (mapcar #'symbol-name (elfeed-entry-tags entry)))
           (tags-str (concat (mapconcat 'identity tags ",")))
           (title-width (- (window-width) elfeed-goodies/feed-source-column-width 4))

           (title-column (elfeed-format-column
                          title (elfeed-clamp
                                 elfeed-search-title-min-width
                                 title-width
                                 elfeed-search-title-max-width)
                          :left))
           (tag-column (elfeed-format-column
                        tags-str (elfeed-clamp elfeed-goodies/tag-column-width
                                               elfeed-goodies/tag-column-width
                                               elfeed-goodies/tag-column-width)
                        :right))
           (feed-column (elfeed-format-column
                         feed-title (elfeed-clamp elfeed-goodies/feed-source-column-width
                                                  elfeed-goodies/feed-source-column-width
                                                  elfeed-goodies/feed-source-column-width)
                         :left)))

      (insert (propertize feed-column 'face 'elfeed-search-feed-face) " ")
      (insert (propertize title 'face title-faces 'kbd-help title))
      (insert (propertize tag-column 'face 'elfeed-search-tag-face) " ")
      (setq-local line-spacing 0.2)))

  (defun +myrss/elfeed-show-refresh--better-style ()
    "Update the buffer to match the selected entry, using a mail-style."
    (interactive)
    (let* ((inhibit-read-only t)
           (title (elfeed-entry-title elfeed-show-entry))
           (date (seconds-to-time (elfeed-entry-date elfeed-show-entry)))
           (author (elfeed-meta elfeed-show-entry :author))
           (link (elfeed-entry-link elfeed-show-entry))
           (tags (elfeed-entry-tags elfeed-show-entry))
           (tagsstr (mapconcat #'symbol-name tags ", "))
           (nicedate (format-time-string "%a, %e %b %Y %T %Z" date))
           (content (elfeed-deref (elfeed-entry-content elfeed-show-entry)))
           (type (elfeed-entry-content-type elfeed-show-entry))
           (feed (elfeed-entry-feed elfeed-show-entry))
           (feed-title (elfeed-feed-title feed))
           (base (and feed (elfeed-compute-base (elfeed-feed-url feed)))))
      (erase-buffer)
      (insert "\n")
      (insert (format "%s\n\n" (propertize title 'face 'elfeed-show-title-face)))
      (insert (format "%s\t" (propertize feed-title 'face 'elfeed-search-feed-face)))
      (when (and author elfeed-show-entry-author)
        (insert (format "%s\n" (propertize author 'face 'elfeed-show-author-face))))
      (insert (format "%s\n\n" (propertize nicedate 'face 'elfeed-log-date-face)))
      (when tags
        (insert (format "%s\n"
                        (propertize tagsstr 'face 'elfeed-search-tag-face))))
      (cl-loop for enclosure in (elfeed-entry-enclosures elfeed-show-entry)
               do (insert (propertize "Enclosure: " 'face 'message-header-name))
               do (elfeed-insert-link (car enclosure))
               do (insert "\n"))
      (insert "\n")
      (if content
          (if (eq type 'html)
              (elfeed-insert-html content base)
            (insert content))
        (insert (propertize "(empty)\n" 'face 'italic)))
      (goto-char (point-min))))
  )
#+END_SRC
*** TODO Tag search
Look into https://gitlab.com/protesilaos/dotfiles/-/blob/master/emacs/.emacs.d/prot-lisp/prot-elfeed.el#L274


doom.d/config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!


;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets.
(setq user-full-name "Abhishek Bansal"
      user-mail-address "abhibansal530@gmail.com")

;; Doom exposes five (optional) variables for controlling fonts in Doom. Here
;; are the three important ones:
;;
;; + `doom-font'
;; + `doom-variable-pitch-font'
;; + `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;;
;; They all accept either a font-spec, font string ("Input Mono-12"), or xlfd
;; font string. You generally only need these two:
;; (setq doom-font (font-spec :family "monospace" :size 14))
;; (setq doom-font (font-spec :family "Source Code Pro" :size 18))
(setq doom-font (font-spec :family "Meslo LG S for Powerline" :size 18))

;; There are two ways to load a theme. Both assume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:
(setq doom-theme 'gruvbox-dark-soft)

;; If you use `org' and don't want your org files in the default location below,
;; change `org-directory'. It must be set before org loads!
(setq org-directory "~/org/")

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type t)


;; Here are some additional functions/macros that could help you configure Doom:
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c g k').
;; This will open documentation for it, including demos of how they are used.
;;
;; You can also try 'gd' (or 'C-c g d') to jump to their definition and see how
;; they are implemented.

;; My custom keybindings
;; (map! :n ";" ":")
;; (define-key evil-normal-state-map ";" ":")
;; (define-key evil-motion-state-map ";" 'evil-ex)

;; (load-theme 'gruvbox-dark-soft-theme t)

;; To apply any custom tag(s) to captured items.
;(defun abhi/tag-inbox-capture ()
;  (when (seq-contains '("c") (plist-get org-capture-plist :key))
;    (org-toggle-tag "PROTOCOL" 'on)))
;
;(add-hook 'org-capture-before-finalize-hook 'abhi/tag-inbox-capture)

(after! org
  (require 'find-lisp)
  (setq org-ditaa-jar-path "/usr/local/Cellar/ditaa/0.11.0_1/libexec/ditaa-0.11.0-standalone.jar")
  (setq org-startup-folded t)
  (setq org-todo-keywords
        '((sequence "TODO(t)" "NEXT(n)" "SOMEDAY(s)" "PROJECT(p)" "|" "DONE(d)")
          (sequence "WAITING(w@/!)" "HOLD(h@/!)" "|" "CANCELLED(c@/!)")))
  (setq abhi/org-agenda-directory "~/org/gtd/")

  ;; Alghout all of these are not required for agenda, but adding these allow for
  ;; searching tags across all these files.
  (setq org-agenda-files (append '("~/org/gtd/inbox.org" "~/org/gtd/life.org" "~/org/gtd/projects.org")
                                 (directory-files-recursively "~/org/resources" "\\.org$")))

  (setq org-complete-tags-always-offer-all-agenda-tags t)

  (setq org-refile-targets '(("~/org/gtd/projects.org" :maxlevel . 1)
                             ("~/org/gtd/life.org" :maxlevel . 3)))

  (setq org-capture-templates
        `(("i" "Inbox")

          ("iP" "Personal inbox" entry
           (file "~/org/gtd/inbox.org")
           (file "~/org/capture-templates/inbox.txt")
           :empty-lines-after 1)

          ("iW" "Work inbox" entry
           (file "~/org/gtd/work/inbox.org")
           (file "~/org/capture-templates/inbox.txt")
           :empty-lines-after 1)

          ("c" "org-protocol-capture" entry
           (file "~/org/gtd/inbox.org")
           (file "~/org/capture-templates/protocol.txt")
           :immediate-finish t)

          ("g" "GTD templates")

          ("gp" "Create a daily plan")

          ("gpP" "Daily plan private" plain
           (file+olp+datetree "~/org/gtd/planner.org")
           (file "~/org/capture-templates/daily-plan.txt")
           :immediate-finish t)

          ("gpW" "Daily plan work" plain
           (file+olp+datetree "~/org/gtd/work/planner.org")
           (file "~/org/capture-templates/daily-plan-work.txt")
           :immediate-finish t)

          ("gr" "Review templates")

          ("grd" "Daily review" plain
           (file+olp+datetree "~/org/gtd/life.org" "GTD" "Daily Review")
           (file "~/org/capture-templates/daily-review.txt")
           :immediate-finish t
           :jump-to-captured t)

          ("grw" "Weekly review" plain
           (file+olp+datetree "~/org/gtd/life.org" "GTD" "Weekly Review")
           (file "~/org/capture-templates/weekly-review.txt")
           :immediate-finish t
           :jump-to-captured t)

          ("r" "Capture a resource")

          ("rB" "Book" entry
           (file "~/org/resources/books.org")
           (file "~/org/capture-templates/book.txt")
           :empty-lines-after 2))))

(setq org-refile-allow-creating-parent-nodes 'confirm)

(add-to-list 'load-path "~/.emacs.d/.local/straight/repos/org-mode/lisp/org-protocol.el")
(require 'org-protocol)

(defun abhi/org-agenda-process-inbox-item ()
  "Process a single item in the org-agenda."
  (org-with-wide-buffer
   (org-agenda-set-tags)
   (org-agenda-refile nil nil t)))

(defun abhi/bulk-process-entries ()
  (if (not (null org-agenda-bulk-marked-entries))
      (let ((entries (reverse org-agenda-bulk-marked-entries))
            (processed 0)
            (skipped 0))
        (dolist (e entries)
          (let ((pos (text-property-any (point-min) (point-max) 'org-hd-marker e)))
            (if (not pos)
                (progn (message "Skipping removed entry at %s" e)
                       (cl-incf skipped))
              (goto-char pos)
              (let (org-loop-over-headlines-in-active-region) (funcall 'abhi/org-agenda-process-inbox-item))
              ;; `post-command-hook' is not run yet.  We make sure any
              ;; pending log note is processed.
              (when (or (memq 'org-add-log-note (default-value 'post-command-hook))
                        (memq 'org-add-log-note post-command-hook))
                (org-add-log-note))
              (cl-incf processed))))
        (org-agenda-redo)
        (unless org-agenda-persistent-marks (org-agenda-bulk-unmark-all))
        (message "Acted on %d entries%s%s"
                 processed
                 (if (= skipped 0)
                     ""
                   (format ", skipped %d (disappeared before their turn)"
                           skipped))
                 (if (not org-agenda-persistent-marks) "" " (kept marked)")))))

(defun abhi/org-process-inbox ()
  "Called in org-agenda-mode, processes all inbox items."
  (interactive)
  (org-agenda-bulk-mark-regexp "new:")
  (abhi/bulk-process-entries))

(map! :map org-agenda-mode-map
      "r" #'abhi/org-process-inbox)

;; Org agenda.
(use-package! org-agenda
  :init
  :config
  (setq org-columns-default-format "%40ITEM(Task) %Effort(EE){:} %CLOCKSUM(Time Spent) %SCHEDULED(Scheduled) %DEADLINE(Deadline)")
  (setq org-agenda-custom-commands `(("c" "Agenda"
                                      ((agenda ""
                                               ((org-agenda-span 'week)
                                                (org-deadline-warning-days 365)))
                                       (todo "TODO"
                                             ((org-agenda-overriding-header "To Refile")
                                              (org-agenda-files '(,(concat abhi/org-agenda-directory "inbox.org")))))
                                       (todo "NEXT"
                                             ((org-agenda-overriding-header "In Progress")
                                              (org-agenda-files '(,(concat abhi/org-agenda-directory "projects.org")
                                                                  ,(concat abhi/org-agenda-directory "next.org")))
                                              ))
                                       (todo "TODO"
                                             ((org-agenda-overriding-header "Projects")
                                              (org-agenda-files '(,(concat abhi/org-agenda-directory "projects.org")))
                                              ))
                                       (todo "TODO"
                                             ((org-agenda-overriding-header "One-off Tasks")
                                              (org-agenda-files '(,(concat abhi/org-agenda-directory "next.org")))
                                              (org-agenda-skip-function '(org-agenda-skip-entry-if 'deadline 'scheduled))))))

                                     ("i" "Inbox"
                                      ((todo ""
                                            ((org-agenda-overriding-header "To Refile")
                                             (org-agenda-files '(,(concat abhi/org-agenda-directory "inbox.org"))))))))))

(use-package! org-roam
  :hook
  (after-init . org-roam-mode)
  :config
  (setq org-roam-ref-capture-templates
        '(("r" "ref" plain (function org-roam-capture--get-point)
           "%?"
           :file-name "${slug}"
           :head "#+TITLE: ${title}
    #+ROAM_KEY: ${ref}
    #+ROAM_TAGS: Website
    - source :: ${ref}"
           :unnarrowed t)))
  (setq +org-roam-open-buffer-on-find-file nil))

(eval-after-load "artist"
  '(define-key artist-mode-map [(down-mouse-3)] 'artist-mouse-choose-operation)
  )

;; (use-package! org-roam-protocol
;;   :after org-protocol)

;; Org jounral.
(use-package org-journal
  :custom
  (org-journal-dir (concat org-roam-directory "/journal"))
  (org-journal-date-prefix "#+TITLE: ")
  (org-journal-file-format "%Y-%m-%d.org")
  (org-journal-date-format "%A, %d %B %Y"))

;; Deft.
(use-package deft
      :after org
      :custom
      (deft-recursive t)
      (deft-use-filter-string-for-filename t)
      (deft-default-extension "org")
      (deft-directory org-roam-directory))

(custom-set-variables
 ;; custom-set-variables was added by Custom.
 ;; If you edit it by hand, you could mess it up, so be careful.
 ;; Your init file should contain only one such instance.
 ;; If there is more than one, they won't work right.
 '(safe-local-variable-values
   (quote
    ((whitespace-line-column . 80)
     (whitespace-style quote
                       (face lines-tail))))))
(custom-set-faces
 ;; custom-set-faces was added by Custom.
 ;; If you edit it by hand, you could mess it up, so be careful.
 ;; Your init file should contain only one such instance.
 ;; If there is more than one, they won't work right.
 )

;; http://www.howardism.org/Technical/Emacs/getting-more-boxes-done.html
;; https://mollermara.com/blog/Fast-refiling-in-org-mode-with-hydras/

(defmacro my-org-make-refile-command (fn-suffix refile-targets)
  "Generate a command to call `org-refile' with modified targets."
  `(defun ,(intern (concat "my-org-refile-" (symbol-name fn-suffix))) ()
     ,(format "`org-refile' to %S" refile-targets)
     (interactive)
     (org-refile-cache-clear)
     (let ((org-refile-target-verify-function nil)
           (org-refile-history nil)
           (org-refile-targets ,refile-targets))
       (if (eq major-mode 'org-agenda-mode)
           (call-interactively 'org-agenda-refile)
         (call-interactively 'org-refile)))))

(my-org-make-refile-command book '(("~/org/resources/books.org" :maxlevel . 2)))
(my-org-make-refile-command course '(("~/org/resources/courses.org" :maxlevel . 1)))
(my-org-make-refile-command link '(("~/org/resources/links.org" :maxlevel . 1)))
(my-org-make-refile-command paper '(("~/org/resources/papers.org" :maxlevel . 1)))
(my-org-make-refile-command read '(("~/org/resources/readings.org" :maxlevel . 1)))
(my-org-make-refile-command watch '(("~/org/resources/watch.org" :maxlevel . 1)))
(my-org-make-refile-command project '(("~/org/gtd/projects.org" :maxlevel . 1)))
(my-org-make-refile-command people '(("~/org/resources/people.org" :maxlevel . 1)))

(defhydra my-org-refile-hydra (:hint nil :foreign-keys run)
  "
^Refile^            ^Goto^                     ^Dired^
------------------------------------------------------
_p_: Projects       _g j_: Last refile         _d r_: Resources
_P_: Papers         _g r_: To Read
_b_: Books          _g p_: Projects
_c_: Courses
_l_: Links
_r_: To Read
_w_: To Watch
_o_: People
_R_: Refile
"
  ("b" my-org-refile-book)
  ("c" my-org-refile-course)
  ("l" my-org-refile-link)
  ("p" my-org-refile-project)
  ("P" my-org-refile-paper)
  ("r" my-org-refile-read)
  ("w" my-org-refile-watch)
  ("o" my-org-refile-people)
  ("R" org-refile)
  ("g j" org-refile-goto-last-stored :exit t)
  ("g r" (find-file-other-window "~/org/resources/readings.org") :exit t)
  ("g p" (find-file-other-window "~/org/gtd/projects.org") :exit t)
  ("d r" (dired "~/org/resources") :exit t)
  ("q" nil "cancel"))

(global-set-key (kbd "<f9> r") 'my-org-refile-hydra/body)

(use-package! nov
  :mode ("\\.epub\\'" . nov-mode)
  :config
  (setq nov-save-place-file (concat doom-cache-dir "nov-places")))
